### ISSUE TITLE
Undoing/redoing changes does not trigger a reblame

### ISSUE THREAD

**Issue by Arnesfield** [mk12/vscode-better-git-line-blame#17]
I've enabled both this and [vscode-git-line-blame](https://github.com/carlthome/vscode-git-line-blame) extensions just to see the differences (also more or less works with [GitLens](https://github.com/gitkraken/vscode-gitlens)).

https://github.com/user-attachments/assets/947fec42-e21c-490f-b5ad-231f5f77e21c

Changing a line then changing it back shows `Uncommitted changes` instead of showing the correct inline annotation message.

By _changing it back_, it's either manually changing the line back to its original contents or via undo/redo.

Currently, a reblame seems to be triggered on a save whenever `betterGitLineBlame.reblameOnSave` is `true`.

Not sure if this is considered an issue or if it needs fixing as the setting `betterGitLineBlame.reblameOnSave` kinda implies (at least to me) that the extension intentionally tries to do the least amount of work when not configured to. But, if it is unexpected behavior, maybe adding another setting to trigger the reblame on update would also work.

**Comment by mk12** [comment]
Yes, I have a FAQ on this: https://github.com/mk12/vscode-better-git-line-blame?tab=readme-ov-file#do-i-need-to-reblame. Also discussed in #3 and #8.

It's never bothered me that much because I can tell from the gutter indicator if a line is unchanged. So I know when I'm in that situation and I can save or manually reblame if I need to. But it's come up many times now so clearly this behavior is surprising people and there will continue to be more bug reports about it.

I'm going to make reblameOnSave true by default in the next release. That should alleviate this somewhat.

I don't know of an easy way to make it update instantly when you revert a change. Some possibilities:

1. Status quo: not instant. Wait until a manual command, git state change, or save event.
2. What [vscode-git-line-blame](https://github.com/carlthome/vscode-git-line-blame) does: spawn a `git blame` process on every keystroke. This feels way too wasteful to me. My extension is fast (even on large repos) because it only blames once. I haven't measured anything, but I suspect it's friendlier to laptop battery life too.
3. What GitLens does. I'm not actually sure what they do, so would have to investigate.
4. Periodically diff to identify no-longer-changed lines and trigger reblame. This would be nice if I could access the diff VS Code has already done for the gutter, but I don't think it's accessible to extensions.
5. Keep track of the original contents of each line such that the blame can always be kept accurate. This is in theory possible, but it would amount to hand writing a diffing algorithm (which might not always match what git does).

**Comment by Arnesfield** [comment]
I played around with it a bit more and I think I'll be using keybindings to trigger the reblame instead.

```json5
// keybindings.json
[
  // undo and redo override, reblame file
  {
    "key": "ctrl+z",
    "command": "runCommands",
    "args": { "commands": ["undo", "betterGitLineBlame.reblameFile"] }
  },
  {
    "key": "ctrl+shift+z",
    "command": "runCommands",
    "args": { "commands": ["redo", "betterGitLineBlame.reblameFile"] }
  },
  // ...
]
```

In the case where a line is changed and then changed back again with its original contents, the file would still need to be saved, so it would reblame after a save anyway.

For the undo/redo case however, since the file wouldn't need to be saved after undoing/redoing the changes, triggering a reblame then should pretty much solve the issue.

What's interesting too is that when using these keybindings, I expected the reblame to get triggered every undo/redo. But for some reason, the reblame does not get triggered until the last undo/redo where the file wouldn't need to be saved, which is ideal and pretty neat.

Maybe adding it to the docs or adding a new setting for it would be good too (like reblame on undo/redo), but I'm currently happy with this solution and consider this issue resolved.

**Comment by mk12** [comment]
> What's interesting too is that when using these keybindings, I expected the reblame to get triggered every undo/redo. But for some reason, the reblame does not get triggered until the last undo/redo where the file wouldn't need to be saved, which is ideal and pretty neat.

This is probably because the extension never runs git blame when the file is dirty. You can see this if you edit a line and then run the Reblame command — the annotation should change to "(Save to blame)".

I'll leave this open for now because I might experiment with ad hoc tracking of the original contents to try and keep the blame more accurate (option 5 in my earlier comment). In the general case where you delete and rewrite the entire file this amounts to implementing a diff algorithm, but I think a simple approach would work well in the majority of cases.
